### PR TITLE
Fast refresh indices to use search shards

### PR DESCRIPTION
**Pull request by kingherc**
The changes of PR #113478 were reverted because it induced ES-8275. Now that the ticket is done, this PR re-introduces the reverted changes.

Fast refresh indices should now behave like non fast refresh indices in how they execute (m)gets and searches. I.e., they should use the search shards.

For BWC, we define a new transport version. We expect search shards to be upgraded first, before promotable shards. Until the cluster is fully upgraded, the promotable shards (whether upgraded or not) will still receive and execute gets/searches locally.

Relates ES-9573

### CODE REVIEW

**Comment by kingherc** [comment]
buildkite test this

File: server/src/main/java/org/elasticsearch/TransportVersions.java [hunk 1]
```diff
@@ -194,6 +194,7 @@ static TransportVersion def(int id) {
     public static final TransportVersion DATA_STREAM_INDEX_VERSION_DEPRECATION_CHECK = def(8_788_00_0);
     public static final TransportVersion ADD_COMPATIBILITY_VERSIONS_TO_NODE_INFO = def(8_789_00_0);
     public static final TransportVersion VERTEX_AI_INPUT_TYPE_ADDED = def(8_790_00_0);
+    public static final TransportVersion FAST_REFRESH_ADAPT = def(8_791_00_0);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/TransportVersions.java:197]: nit: I would probably prefer to call this `FAST_REFRESH_RCO_2` just to signal that this is the successor of the prior transport version.

**Review by henningandersen — COMMENTED** [review]
Besides the new transport version is this then a clean reintroduction of the original change or did you have to adapt it?

File: server/src/main/java/org/elasticsearch/TransportVersions.java (hunk 1 above)
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/TransportVersions.java]: Done.

**Review by kingherc — COMMENTED** [review]
> Besides the new transport version is this then a clean reintroduction of the original change or did you have to adapt it?

@henningandersen clean re-introduction. Only the part around BitsetFilterCache.java is different than the original PR, because we did not revert the changes in that one (meaning, we left it fixed rather than reverting it to the wrong older code). So now it's a more straightforward change to adapt fast refresh there.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**